include/revng/StackAnalysis/StackAnalysis.h

```diff
@@ -52,4 +62,169 @@ class StackAnalysis : public llvm::ModulePass {
+    FakeFunction(FakeFunction) {}
+
+  FuncSummary() = delete;
+  FuncSummary(const FuncSummary &) = delete;
+  FuncSummary(FuncSummary &&) = default;
+  FuncSummary &operator=(const FuncSummary &) = delete;
+  FuncSummary &operator=(FuncSummary &&) = default;
+
+  model::FunctionType::Values Type;
+  std::set<llvm::GlobalVariable *> ClobberedRegisters;
+  llvm::SmallSet<std::pair<llvm::CallInst *, model::FunctionEdgeType::Values>,
+                 4>
+    Result;
+  std::optional<uint64_t> ElectedFSO;
+  llvm::Function *FakeFunction;
+
+  static bool compare(const FuncSummary &Old, const FuncSummary &New) {
+    if (New.Type == Old.Type)
+      return std::includes(Old.ClobberedRegisters.begin(),
+                           Old.ClobberedRegisters.end(),
+                           New.ClobberedRegisters.begin(),
+                           New.ClobberedRegisters.end());
+
+    return (New.Type <= Old.Type);
+  }
+
+  void dump() const debug_function { dump(dbg); }
+
+  template<typename T>
+  void dump(T &Output) const {
+    Output << "Dumping summary of the function. \n"
+           << "Type: " << Type << "\n"
+           << "Clobbered registers: ";
+
+    for (auto *Reg : ClobberedRegisters)
+      Output << ::getName(Reg);
+    Output << "\n";
+  }
+};
+
+class FunctionProperties {
+private:
+  /// \brief Map from CFEP to its function description
+  std::map<llvm::BasicBlock *, FuncSummary> Bucket;
+  FuncSummary DefaultSummary;
+
+  const FuncSummary &get(llvm::BasicBlock *BB) const {
+    auto It = Bucket.find(BB);
+    if (It != Bucket.end())
+      return It->second;
+    return DefaultSummary;
+  }
+
+public:
+  FunctionProperties(FuncSummary DefaultSummary) :
+    DefaultSummary(std::move(DefaultSummary)) {}
+
+  model::FunctionType::Values getFunctionType(llvm::BasicBlock *BB) const {
+    return get(BB).Type;
+  }
+
+  bool isFakeFunction(llvm::BasicBlock *BB) const {
+    return getFunctionType(BB) == model::FunctionType::Values::Fake;
+  }
+
+  llvm::Function *getFakeFunction(llvm::BasicBlock *BB) const {
+    return get(BB).FakeFunction;
+  }
+
+  const auto &getRegistersClobbered(llvm::BasicBlock *BB) const {
+    return get(BB).ClobberedRegisters;
+  }
+
+  std::optional<uint64_t> getElectedFSO(llvm::BasicBlock *BB) const {
+    return get(BB).ElectedFSO;
+  }
+
+  bool registerFunc(llvm::BasicBlock *BB, FuncSummary &&F) {
+    auto It = Bucket.find(BB);
+    if (It != Bucket.end()) {
+      bool Changed = FuncSummary::compare(It->second, F);
+      It->second = std::move(F);
+      return !(Changed);
+    }
+    Bucket.try_emplace(BB, std::move(F));
+    return true;
+  }
+};
+
+template<class FunctionOracle>
+class CFEPAnalyzer {
+  llvm::Module &M;
+  llvm::LLVMContext &Context;
+  GeneratedCodeBasicInfo *GCBI;
+  FunctionOracle &Oracle;
+  llvm::Function *PreHookOpqF;
+  llvm::Function *PostHookOpqF;
+  llvm::Function *IndirectBranchInfoOpqF;
+  OpaqueFunctionsPool<llvm::StringRef> OFPRegistersClobbered;
```

Comment:
Do not prefix the field name with its type. Say no to [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation).

Reply:
Who said it? Is this a revng convention?

Reply:
Not everything is encoded in the coding conventions document. In particular it's hard to make a list of all the things *not* to do.
But I'll add it.

Reply:
Considering that the Hungarian notation is still widely used nowadays, perhaps it would've been better to know it. But I changed them all.